File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/flatmm_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/flatmm_emit.c -- C-side emitter for the FlatMM parity harness.
 * Selects one of 6 sampled FlatMMSpec configs by argv[1] (0..5), builds
 * rocke_flatmm_spec_t identically to the Python emitter flatmm_emit.py, builds
 * the kernel via rocke_build_flatmm and lowers via rocke_lower_kernel_to_llvm
 * (arch gfx950, flavor AUTO), and prints the .ll to stdout so the two outputs
 * can be byte-compared.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_flatmm.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, rocke_flatmm_spec_t* spec)
{
    *spec = rocke_flatmm_spec_default();
    spec->name = "rocke_flatmm";
    spec->wave_size = 64;
    spec->batch_size = 0;
    spec->preshuffle_b = false;

    switch(idx)
    {
    case 0:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 128,
                                              .tile_n = 128,
                                              .tile_k = 64,
                                              .warp_m = 1,
                                              .warp_n = 4,
                                              .warp_k = 1,
                                              .warp_tile_m = 32,
                                              .warp_tile_n = 32,
                                              .warp_tile_k = 16};
        spec->trait.pipeline = "compv4";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "cshuffle";
        spec->block_size = 256;
        break;
    case 1:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 128,
                                              .tile_n = 128,
                                              .tile_k = 64,
                                              .warp_m = 1,
                                              .warp_n = 4,
                                              .warp_k = 1,
                                              .warp_tile_m = 32,
                                              .warp_tile_n = 32,
                                              .warp_tile_k = 16};
        spec->trait.pipeline = "mem";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "default";
        spec->block_size = 256;
        break;
    case 2:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 128,
                                              .tile_n = 128,
                                              .tile_k = 64,
                                              .warp_m = 1,
                                              .warp_n = 4,
                                              .warp_k = 1,
                                              .warp_tile_m = 16,
                                              .warp_tile_n = 16,
                                              .warp_tile_k = 32};
        spec->trait.pipeline = "compv4";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "cshuffle";
        spec->block_size = 256;
        break;
    case 3:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 64,
                                              .tile_n = 64,
                                              .tile_k = 32,
                                              .warp_m = 1,
                                              .warp_n = 2,
                                              .warp_k = 1,
                                              .warp_tile_m = 32,
                                              .warp_tile_n = 32,
                                              .warp_tile_k = 16};
        spec->trait.pipeline = "compv4";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "cshuffle";
        spec->block_size = 128;
        break;
    case 4:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 128,
                                              .tile_n = 128,
                                              .tile_k = 64,
                                              .warp_m = 2,
                                              .warp_n = 2,
                                              .warp_k = 1,
                                              .warp_tile_m = 32,
                                              .warp_tile_n = 32,
                                              .warp_tile_k = 16};
        spec->trait.pipeline = "compv4";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "cshuffle";
        spec->block_size = 512;
        break;
    case 5:
        spec->tile = (rocke_gemm_tile_spec_t){.tile_m = 256,
                                              .tile_n = 256,
                                              .tile_k = 64,
                                              .warp_m = 2,
                                              .warp_n = 4,
                                              .warp_k = 1,
                                              .warp_tile_m = 32,
                                              .warp_tile_n = 32,
                                              .warp_tile_k = 16};
        spec->trait.pipeline = "compv4";
        spec->trait.scheduler = "intrawave";
        spec->trait.epilogue = "cshuffle";
        spec->block_size = 512;
        break;
    default:
        return -1;
    }
    rocke_flatmm_spec_finalize(spec);
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index 0..5>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    rocke_flatmm_spec_t spec;
    if(make_spec(idx, &spec) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    rocke_ir_builder_t b;
    char namebuf[256];
    if(rocke_flatmm_kernel_name(&spec, namebuf, sizeof namebuf) != ROCKE_OK)
    {
        fprintf(stderr, "kernel_name failed\n");
        return 1;
    }
    rocke_ir_builder_init(&b, namebuf);

    rocke_kernel_def_t* kernel = rocke_build_flatmm(&b, &spec, "gfx950");
    if(!kernel)
    {
        fprintf(stderr, "build_flatmm failed: %s\n", rocke_ir_builder_error(&b));
        rocke_ir_builder_free(&b);
        return 1;
    }

    int ret = 0;
    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, "gfx950", &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* t = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &t);
        if(st != ROCKE_OK || !t)
        {
            fprintf(stderr, "ir_serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
    }
    else if(strcmp(mode, "verify") == 0)
    {
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }
    else
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        rocke_ir_builder_free(&b);
        return 2;
    }
    rocke_ir_builder_free(&b);
    return ret;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/flatmm_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/flatmm_emit.py -- Python reference emitter for the FlatMM
# parity harness. Selects one of 6 sampled FlatMMSpec configs by argv[1]
# (0..5), builds it via build_flatmm and prints lower_kernel_to_llvm(
# arch='gfx950') to stdout so it can be byte-compared with the C emitter
# flatmm_emit.c.
from rocke.instances.common.flatmm import FlatMMSpec, build_flatmm
from rocke.instances.common.gemm_universal import TileSpec, TraitSpec
from _emit_common import run_emit


def _spec(idx: int) -> FlatMMSpec:
    if idx == 0:
        return FlatMMSpec(
            tile=TileSpec(
                128,
                128,
                64,
                warp_m=1,
                warp_n=4,
                warp_k=1,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
            ),
            trait=TraitSpec(
                pipeline="compv4", scheduler="intrawave", epilogue="cshuffle"
            ),
            wave_size=64,
            block_size=256,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    if idx == 1:
        return FlatMMSpec(
            tile=TileSpec(
                128,
                128,
                64,
                warp_m=1,
                warp_n=4,
                warp_k=1,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
            ),
            trait=TraitSpec(pipeline="mem", scheduler="intrawave", epilogue="default"),
            wave_size=64,
            block_size=256,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    if idx == 2:
        return FlatMMSpec(
            tile=TileSpec(
                128,
                128,
                64,
                warp_m=1,
                warp_n=4,
                warp_k=1,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=32,
            ),
            trait=TraitSpec(
                pipeline="compv4", scheduler="intrawave", epilogue="cshuffle"
            ),
            wave_size=64,
            block_size=256,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    if idx == 3:
        return FlatMMSpec(
            tile=TileSpec(
                64,
                64,
                32,
                warp_m=1,
                warp_n=2,
                warp_k=1,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
            ),
            trait=TraitSpec(
                pipeline="compv4", scheduler="intrawave", epilogue="cshuffle"
            ),
            wave_size=64,
            block_size=128,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    if idx == 4:
        return FlatMMSpec(
            tile=TileSpec(
                128,
                128,
                64,
                warp_m=2,
                warp_n=2,
                warp_k=1,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
            ),
            trait=TraitSpec(
                pipeline="compv4", scheduler="intrawave", epilogue="cshuffle"
            ),
            wave_size=64,
            block_size=512,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    if idx == 5:
        return FlatMMSpec(
            tile=TileSpec(
                256,
                256,
                64,
                warp_m=2,
                warp_n=4,
                warp_k=1,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
            ),
            trait=TraitSpec(
                pipeline="compv4", scheduler="intrawave", epilogue="cshuffle"
            ),
            wave_size=64,
            block_size=512,
            batch_size=0,
            preshuffle_b=False,
            name="rocke_flatmm",
        )
    raise SystemExit(f"unknown config index {idx}")


def main() -> int:
    return run_emit(
        _spec,
        build_flatmm,
        usage="usage: flatmm_emit.py <config_index 0..5>\n",
    )


if __name__ == "__main__":
    raise SystemExit(main())
